
**Fall 2019**

**P556: Applied Machine Learning**

**Assignment #1**

**Due date: September 18, 2019. 11:59 PM**

DO NOT CHANGE THE FUNCTION DEFINITIONS UNLESS APPROVED BY AN AI.

# Problem #1: Linear Regression

##  Problem 1.1 (25 points)

Implement linear regression using gradient descent. Your implementation should be able to handle simple and multiple linear regression.

Note 1: by implementation we mean that everything has to be written from scratch and that you cannot call a linear regression function from a library, such as sklearn. Usage of standard libraries, such as numpy, pandas, etc., is fine. If you are unsure about whether a library can be used, please contact the AIs well in advance of the submission date.

Note 2: You are free to use sklearn to test whether your results match that from a battle-tested library. This is a great way to know before hand whether your submission is correct. Make sure to use the same parameters on both models before you spend an eternity debugging code that is correct but not returning the same values.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import load_boston
ds=load_boston()

dfx = pd.DataFrame(ds.data, columns=ds.feature_names)
dfy = pd.DataFrame(ds.target,columns=["target"])

# defining feature matrix(X) and response vector(Y)
X=dfx
Y=dfy
X = (X - np.mean(X))
X=X/np.std(X)
#print(X.shape)


class linear_regression:
    def __init__(self, learning_rate=0.05, iterations=1000, fit_intercept=True, normalize=False, coef=None):
        self.fit_intercept = fit_intercept
        self.normalize = normalize
        self.learning_rate = learning_rate
        self.iterations = iterations
        self.coef = coef
        
    def predict(self, X):

        Y_predict = np.dot(X,self.fit_intercept) 
        return Y_predict
    
    def fit(self, X, Y):
        m=len(Y)
        self.fit_intercept = np.zeros([X.shape[1]+1,1])  #size 14*1
        ones = np.ones((m,1))  #size 506
        X = np.hstack((ones, X))
        self.cost=[]
        Y_predict = self.predict(X) - Y
        cost1=np.sum(np.power((Y_predict), 2)) / (2*m)
        print('The estimated cost before updation',cost1)
        
        for i in range(self.iterations):
            Y_predict = self.predict(X) - Y
            error = np.dot(X.T, Y_predict)
            self.fit_intercept = self.fit_intercept - (self.learning_rate/m) * error
            costcal=np.sum(np.power((Y_predict), 2)) / (2*m)
            self.cost.append(costcal)
        print('Cost after updation is:',costcal)
        print('Regression coefficients:',self.fit_intercept)
        return self
        
lr=linear_regression()
lr.fit(X,Y)



The estimated cost before updation target    1184.293834
dtype: float64
Cost after updation is: target    328.946928
dtype: float64
Regression coefficients: [[ 2.25328063e+01]
 [-9.21625901e-01]
 [ 1.07012433e+00]
 [ 1.05613181e-01]
 [ 6.86780128e-01]
 [-2.05010502e+00]
 [ 2.68072587e+00]
 [ 1.39686204e-02]
 [-3.10628267e+00]
 [ 2.57318542e+00]
 [-1.97576133e+00]
 [-2.05721328e+00]
 [ 8.48667146e-01]
 [-3.74018384e+00]]


## Problem 1.2 (10 points)

- Split the Boston Housing dataset into train and test sets (70% and 30%, respectively) (5 points). 
- Fit your linear regression implementation using the training set and print your model's coefficients. Make predictions for the test set using your fitted model (5 points).

In [0]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import load_boston
from sklearn.model_selection  import train_test_split
ds=load_boston()

dfx = pd.DataFrame(ds.data, columns=ds.feature_names)
dfy = pd.DataFrame(ds.target,columns=["target"])

# defining feature matrix(X) and response vector(Y)
X=dfx
Y=dfy
X = (X - np.mean(X))
X=X/np.std(X)

class linear_regression:
    def __init__(self, learning_rate=0.05, iterations=1000, fit_intercept=True, normalize=False, coef=None):
        self.fit_intercept = fit_intercept
        self.normalize = normalize
        self.learning_rate = learning_rate
        self.iterations = iterations
        self.coef = coef
        
    def predict(self, X):

        Y_predict = np.dot(X,self.fit_intercept) 
        return Y_predict
    
    def fit(self, X, Y):
        X = (X - np.mean(X))
        X=X/np.std(X)
        m=len(Y)
        self.fit_intercept = np.zeros([X.shape[1]+1,1])  #size 14*1
        ones = np.ones((m,1))  #size 506
        X = np.hstack((ones, X))#size 506*1
        self.cost=[]
        Y_predict = self.predict(X) - Y
        cost1=np.sum(np.power((Y_predict), 2)) / (2*m)
        print('The estimated cost before updation',cost1)
        
        for i in range(self.iterations):
            Y_predict = self.predict(X) - Y
            error = np.dot(X.T, Y_predict)
            self.fit_intercept = self.fit_intercept - (self.learning_rate/m) * error
            costcal=np.sum(np.power((Y_predict), 2)) / (2*m)
            self.cost.append(costcal)
        print('Cost after updation is:',costcal)
        print('Regression coefficients:',self.fit_intercept)
        return self

trainset_X, testset_X, trainset_Y, testset_Y=train_test_split(dfx,dfy,test_size=0.30)
print("Input traning set size: "+str(trainset_X.shape))
print("Input test set size: "+str(testset_X.shape))
print("Output traning set size: "+str(trainset_Y.shape))
print("Output test set size: "+str(testset_Y.shape))
print("\nDataset split successful")


#print(trainset_X)#354*13
#print(trainset_Y)#354*1

lr=linear_regression()
lr.fit(trainset_X, trainset_Y)


## Problem 1.3 (10 points)

Identify the variable or set of variables that will minimize the mean square error (MSE). Hint: this is where your function being able to handle simple and multiple regression becomes useful.

In [ ]:
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import load_boston
from sklearn.model_selection  import train_test_split
ds=load_boston()

dfx = pd.DataFrame(ds.data, columns=ds.feature_names)
dfy = pd.DataFrame(ds.target,columns=["target"])

# defining feature matrix(X) and response vector(Y)
X=dfx
Y=dfy
X = (X - np.mean(X))
X=X/np.std(X)

class linear_regression:
    def __init__(self, learning_rate=0.05, iterations=1000, fit_intercept=True, normalize=False, coef=None):
        self.fit_intercept = fit_intercept
        self.normalize = normalize
        self.learning_rate = learning_rate
        self.iterations = iterations
        self.coef = coef
        
    def predict(self, X):

        Y_predict = np.dot(X,self.fit_intercept) 
        return Y_predict
    
    def fit(self, X, Y):
        X = (X - np.mean(X))
        X=X/np.std(X)
        m=len(Y)
        self.fit_intercept = np.zeros([X.shape[1]+1,1])  #size 14*1
        ones = np.ones((m,1))  #size 506
        X = np.hstack((ones, X))#size 506*1
        self.mse=[]
        Y_predict = self.predict(X) - Y
        cost1=np.sum(np.power((Y_predict), 2)) / (2*m)
        print('The estimated cost before updation',cost1)
        
        for i in range(self.iterations):
            Y_predict = self.predict(X) - Y
            error = np.dot(X.T, Y_predict)
            self.fit_intercept = self.fit_intercept - (self.learning_rate/m) * error
            mean_error=(1/m)*np.sum((Y_predict.T), 2)
            self.cost.append(mse)
       
       
        print('Mean square error:',self.mse)
        return self



#print(trainset_X)#354*13
#print(trainset_Y)#354*1

lr=linear_regression()




# Your code goes here

## Problem 1.4 (5 points)

1. How do you interpret that a variable causes a model's mean square error to increase? (2 points)
  - Answer: 
  Individually if I evaluate the function for the input data set with individual parameters at a time, the output error would vary. For those parameters that have the most error value ,ie, dfifference between the actual and predicted values, gives the variable that causes the most mean square error. For all the cases where the error squared difference is most are the parameters causing the most noise in the tested system.
  Checking individually y=beta*x+c for parametrs beta gives the error prediction and eventually the mse of that dataset.
2. Why we would want to normalize our variables? (1 point)
  - Answer:
  When the range of features does not differ uniformly, as in ,if some range from 0 to 100 and the orher one from 0 to 1000, we need to scale down the larger parameters to bring the values closr to mean and thereby reduce the deviation. When we do not  normalize, the elements form an elliptical path around the minimum position. When we scale down the values , the plot turns into a concentric circles almost equally deviated from mean each side.The path to the minimum is easier to traverse in such case.This makes the convergence fater.
3. A model fitted using the exact same split dataset with normalized values will generate the same coefficients as a model that was fitted using values that haven't been normalized. Clearly state whether that statement is true or false and explain your reasoning. (2 points)
  - Answer:
 Statement is not completely true.In most cases the coefficients remain the same rather than varying too much. Not normalizing makes the data space a not so smooth function giving haphazard coefficients . Also training the set would take too much time to reach the global minimum if the dataset is not normalized. Also the proportion of weights of the train and test dataset remain the same resulting in similar value coefficients. 

# Problem 2: Binary Classification

## Problem 2.1 (5 points)

Consider the binary classification problem of mapping a given input to two classes. Let $\mathcal{X}=\mathbb{R}^d$ and $\mathcal{Y}=\{+1, -1\}$ be the input space and output space, respectively. In simple words, it means that the input has $d$ features and all of them are real valued, whereas the output can only take values $-1$ or $+1$. This is one of the most common problems in machine learning and many sophisticated methods exist to solve it. In the question, we will solve it using the concepts we have already learned in class. Let us assume the two sets of points can be separated using a straight line i.e. the samples are linearly separable. So if $d=2$, one should be able to draw a line to distinguish between the two classes. All points lying on side of the line should belong to a particular class (say $1$) and the points lying on the other side should belong to another class (say $2$). To see what this would look like,  your first task is as follows:

Write a function that will randomly generate a dataset for this problem. Your function should randomly choose a line $l$, which can be denoted as $ax + by + c = 0$. According to basic high school geometry, the line divides the plane into two sides. On one side, $ax+by+c>0$ while on the other $ax+by+c<0$. Use this fact to randomly generate $k_0$ points on the side of class 0 (i.e. $y=-1$) and $k_1$ points on the side of class 1 (i.e. $y=1$). Create a plot of this dataset where all the points corresponding to one class are blue and those of the other class are green, the line dividing both classes should be red. Axes should be labeled.

**Note**: Do not confuse the $x$ and $y$ in the equation of line $ax + by + c = 0$ with $\mathcal{X} $ and $\mathcal{Y}$. Instead imagine these $x$ and $y$ as the 2-D coordinate system on which you have different points which should lie on 2 sides of the line $ax + by + c = 0$. For example, there is a point (2,3) in the 2-D system where $x = 2$ and $y = 3$.

In [0]:
import random
import matplotlib.pyplot as plt
import numpy as np
def generate_dataset(k0, k1):
    """
    Parameters
    ----------
    k0 : integer, number of samples for class 0
    k1 : integer, number of samples for class 1
    
    Returns
    -------
    X : array, shape (m, d), dimension numpy array where m is the number of 
    samples and d is the number of features 

    Y : array, (m, 1), dimension vector where m is the number of samples
    """
    a=random.uniform(-10,10)
    print('a:',a)
    num_points=10
    b=random.uniform(-10,10)
    print('b:',b)
    c=random.uniform(-10,10)
    print('c:',c)
    
    x=random.sample(range(-30,30), 10)
    
    y=[]
    for i in range(len(x)):
        
        y.append(-((a*x[i])+c)/b)
        
    plt.plot(x,y,color = 'red')
    
    data,target=[],[]
    
    k0x, k0y = [],[]
    while len(data) <= k0:
        x, y = random.uniform(-20,20), random.uniform(-20,20)
        if (a*x)+(b*y)+c < 0: 
            k0x.append(x) 
            k0y.append(y)
            data.append([x,y])
            target.append(-1)
                
    k1x, k1y = [],[]
    while len(data) <= (k0+k1):
        x, y = random.uniform(-20,20), random.uniform(-20,20)
        if (a*x)+(b*y)+c > 0:
            k1x.append(x)
            k1y.append(y)
            data.append([x,y])
            target.append(1)
            
    plt.scatter(k0x,k0y,color ='b',label='-1')
    plt.scatter(k1x,k1y,color ='g',label='1')
    plt.xlabel('x-Axis')
    plt.ylabel('y-Axis')
    plt.show()
    
    
    return np.array(data), np.array(target)

X,Y = generate_dataset(30,15)
    pass

## Problem 2.2 (35 points)

If $\mathcal{Y}$ is the variable you are trying to predict using a feature $\mathcal{X}$ then in a typical Machine Learning problem, you are tasked with a target function $f$ which maps $\mathcal{X}$ to $\mathcal{Y}$ i.e. Find $f$ such that  $\mathcal{Y}$  = $f(\mathcal{X})$


When you are given a dataset for which you do not have access the target function $f$, you have to learn it from the data. In this problem, we are going to learn the parameters of the line that separates the two classes for the dataset that we constructed in Problem 2.1. As we previously mentioned, that line can be represented as $ax + by + c = 0$.

The goal here is to correctly find out the coefficients $a$, $b$, and $c$, represented below as $\bf{w}$ which is a vector. The algorithm to find it is a simple iterative process: 

1. Randomly choose a $\mathbf{w}$ to begin with.
2. Keep on adjusting the value of $\bf{w}$ as follows until all data samples are correctly classified:
    1. Randomly choose a sample from the dataset without replacement and see if it is correctly classified. If yes,  move on to another sample.
    2. If not,  then  update the weights as $\mathbf{w}^{t+1} = \mathbf{w}^t + y \cdot \mathbf{x}$
    and go back to the previous step (of randomly chosing a sample)
    
        - $\mathbf{w}^{t+1}$ is value of $\mathbf{w}$ at iteration $t+1$
        - $\mathbf{w}^{t}$ is value of $\mathbf{w}$ at iteration $t$
        - $y$ is the class label for the sample under consideration
        - $\mathbf{x}$ is the data-point under consideration
    
    
Write a function that implements this learning algorithm. The input to the function is going to be a dataset represented by the input variable $X$ and the target variable $y$. The output of the function should be the chosen $\mathbf{w}$.



In [0]:
import random
import matplotlib.pyplot as plt
import numpy as np
def generate_dataset(k0, k1):
    """
    Parameters
    ----------
    k0 : integer, number of samples for class 0
    k1 : integer, number of samples for class 1
    
    Returns
    -------
    X : array, shape (m, d), dimension numpy array where m is the number of 
    samples and d is the number of features 

    Y : array, (m, 1), dimension vector where m is the number of samples
    """
    a=random.uniform(-10,10)
    print('a:',a)
    num_points=10
    b=random.uniform(-10,10)
    print('b:',b)
    c=random.uniform(-10,10)
    print('c:',c)
    
    x=random.sample(range(-30,30), 10)
    
    y=[]
    for i in range(len(x)):
        
        y.append(-((a*x[i])+c)/b)
        
    plt.plot(x,y,color = 'red')
    
    data,target=[],[]
    
    k0x, k0y = [],[]
    while len(data) <= k0:
        x, y = random.uniform(-20,20), random.uniform(-20,20)
        if (a*x)+(b*y)+c < 0: 
            k0x.append(x) 
            k0y.append(y)
            data.append([x,y])
            target.append(-1)
                
    k1x, k1y = [],[]
    while len(data) <= (k0+k1):
        x, y = random.uniform(-20,20), random.uniform(-20,20)
        if (a*x)+(b*y)+c > 0:
            k1x.append(x)
            k1y.append(y)
            data.append([x,y])
            target.append(1)
            
    plt.scatter(k0x,k0y,color ='b',label='-1')
    plt.scatter(k1x,k1y,color ='g',label='1')
    plt.xlabel('x-Axis')
    plt.ylabel('y-Axis')
    plt.show()
    
    
    return np.array(data), np.array(target)

X,Y = generate_dataset(30,15)
#print(X.shape) #46*2   
#M=len(X[0])+1#3
#M

def fit_line(X, y):
    """Predict using the binary classification model. Use the dataset generated 
    using generate_data() as input for this function.
    
    Parameters
    ----------
    X : array_like, shape (n_samples, n_features)
        Samples.
    y : array_like, shape (n_labels, 1)
    
    Returns
    -------
    w : array, shape (1,n_features)
        Returns the final weight vector w.  
    """
    w=np.array([0.0] * (1+len(X[0])))
    ones = np.ones((len(X),1))
    X = np.hstack((ones, X))
    
    lowerbound_x, lowerbound_y, upperbound_x, upperbound_y = [], [], [], []
    
    for i in range(len(X)):
        if Y[i] == -1: 
            lowerbound_x.append(X[:,1][i]) 
            lowerbound_y.append(X[:,2][i])

        if Y[i] == 1: 
            upperbound_x.append(X[:,1][i]) 
            upperbound_y.append(X[:,2][i])

        if np.dot(X[i], w.T) != Y[i]:
            m=Y[i] * X[i]
            w = w + m
            
    X_plot, Y_plot = X[:,1], []
    
    Y_plot=[]
    for i in range(len(X)):
        Y_plot.append(-(w[0] + (w[1] * X[i][1]))/w[2])
    
    plt.scatter(lowerbound_x, lowerbound_y, color = 'blue',label='-1') 
    plt.scatter(upperbound_x, upperbound_y, color = 'green',label='1')
    plt.xlabel('X-Axis'),plt.ylabel('Y-Axis')
    plt.plot(X_plot, Y_plot, color = 'red')
    print("The corresponding weights are %s" %(w))
    return w

fit_line(X,Y)
    
            


### Problem 2.3 (10 points)
- Give an intuition of why the above algorithm converges for linearly separable data? We do not expect you to give a mathematic proof, but it would be great if you can provide one. You will get full points even if you just provide an intuition of a few lines. Including figures or mathematical equations is encouraged but not required. (5 points)

  - Answer: 
  The above algorithm best fit for the generated dataset by taking the estimated values of y calculated using the parameters and data matrix. Then it takes these values to calculate the lower and upper bounds for the line. Eventually based on these results it plots a separation line. This line gets updated to accurately fit the points by calculating the product of transpose(vector w) and X rows to get the bias line corrected to the accurate position. 

- What happens when the data is not linearly separable? What can be done to salvage the situation?

  - Answer:
  The function we used to predict best fit line only linearly classifies the data. But this is not useful to describe/classify the data in an accurate way. The data points at or near the junction get linearly classified into the other class if they fall above the best fit line. For describing the data accurately, we need a curvy separation for exactly dividing the classes separetly. We cannot determine non linear data with a straight line function.
  When the data is non-linearly separated, there is a high chance that the points get merged into other classes. For this type of classification, we can no longer use the linear separation line.
  For solving such type of equations, we need to use higher degree polynomial regression to classify the data clusters more accurately.